### PR TITLE
Babel UI initial commit

### PR DESCRIPTION
**Pull request by matiascalegaris**
Add babel dependencies
Split all logic requiered from connect forms se we can have both babel and vb6 form working together using user settings

this PR is the first step to integrate a full html5 UI with vb6

### CODE REVIEW

File: CODIGO/BabelUI.bas [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+Private Type BABELSETTINGS
+    Width As Long
+    Height As Long
+    Compresed As Long
+    EnableDebug As Long
+End Type
+
+Private ServerEnvironment As String
+
+Public Declare Function InitializeBabel Lib "BabelUI.dll" Alias "_InitializeBabel@4" (ByRef Settings As BABELSETTINGS) As Boolean
```
Comment (morgolock) [inline CODIGO/BabelUI.bas:66]: Why `@4` in `_InitializeBabel@4` ? the same for all the other public functions in the dll, it should not be necessary to use `@x`

File: CODIGO/BabelUI.bas [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+        GetMainHwdn = frmConnect.hwnd
+    End If
+    
+End Function
+
+Public Sub InitializeUI(ByVal Width As Long, ByVal Height As Long, ByVal pixelSize As Long)
+On Error GoTo InitializeUI_Err
+    If BabelInitialized Then Exit Sub
+100 Dim initSuccess As Boolean
+102 UITexture.Height = Height
```
Comment (morgolock) [inline CODIGO/BabelUI.bas:154]: debug.assert width > 0 and height > 0 and pixelsize > 0

File: CODIGO/BabelUI.bas (hunk 1 above)
Comment (matiascalegaris) [reply CODIGO/BabelUI.bas:66]: I was not able to configure a different calling convention that works with vb6 not sure why

File: CODIGO/BabelUI.bas [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+116 Set UITexture.Texture = SurfaceDB.CreateTexture(UITexture.TextureWidth, UITexture.TextureHeight)
+118 BabelInitialized = True
+    Call RegisterCallbacks(AddressOf LoginCB, AddressOf CloseClientCB, AddressOf BabelUI.CreateAccount, AddressOf SetHostCB, AddressOf ValidateCodeCB, AddressOf ResendValidationCodeCB, AddressOf RequestPasswordResetCB, AddressOf RequestNewPasswordCB, AddressOf SelectCharacterPreviewCB, AddressOf LoginCharacterCB, AddressOf ReturnToLoginCB, AddressOf CreateCharacterCB)
+    Exit Sub
+InitializeUI_Err:
+    Call RegistrarError(Err.Number, Err.Description, "BabelUI.InitializeUI", Erl)
+End Sub
+
+Public Sub InitializeInspectorUI(ByVal Width As Long, ByVal Height As Long)
+On Error GoTo InitializeInspectorUI_Err
```
Comment (morgolock) [inline CODIGO/BabelUI.bas:179]: `debug.assert width > 0 and height > 0 and pixelsize > 0`

File: CODIGO/BabelUI.bas [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+    Exit Sub
+InitializeInspectorUI_Err:
+    Call RegistrarError(Err.Number, Err.Description, "BabelUI.InitializeInspectorUI", Erl)
+End Sub
+
+Public Sub DrawUITexture(ByRef TextureInfo As t_UITexture)
+On Error GoTo DrawTexture_Err
+    With TextureInfo
+100     If Not .Texture Is Nothing Then
+110         If .Texture Is Nothing Then
```
Comment (morgolock) [inline CODIGO/BabelUI.bas:203]:
```
 If Not .Texture Is Nothing Then
   If .Texture Is Nothing Then
```
The second if can be removed, surely if the first is true the second one cannot be true?

File: CODIGO/BabelUI.bas (hunk 1 above)
Comment (morgolock) [reply CODIGO/BabelUI.bas:66]: Ok we need to fix it, I have this solved in `steam_vb.dll` its a matter of using the same configuration

Comment (matiascalegaris) [reply CODIGO/BabelUI.bas:66]: lets have a call to look at this then I tried multiple settings but when I exposed other calling conventions VB6 could not find the simbols

File: CODIGO/General.bas [hunk 5]
```diff
@@ -1133,25 +1133,22 @@ On Error GoTo Main_Err
     CheckMD5 = GetMd5
     SessionOpened = False
     
-
-    Call Load(frmConnect)
-    Call Load(FrmLogear)
-        
+    If CheckAndSetBabelUIUsage Then
+        Call InitializeUI(D3DWindow.BackBufferWidth, D3DWindow.BackBufferHeight, 4)
```
Comment (morgolock) [inline CODIGO/General.bas:1137]: `4` ? I would use a const named variable instead of just `4`

Comment (morgolock) [inline CODIGO/General.bas:1137]: Also we should print a message `debug.print Babel Failed` if the initialization fails for any reason

File: CODIGO/ModLogin.bas [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+                    'When not in DEVELOPER mode we read the ip and port from the list
+                    Call SetDefaultServer
+                #Else
+                    'Staging, set the ip and port for pymmo
+                    IPdelServidorLogin = "45.235.98.31"
+                    PuertoDelServidorLogin = 11814
+                #End If
+            #End If
+        #End If
+    End If
```
Comment (morgolock) [inline CODIGO/ModLogin.bas:65]: At the end of the function please add

```
Debug.Print "Using Login Server " & IPdelServidorLogin & ":" & PuertoDelServidorLogin
Debug.Print "Using Game Server " & IPdelServidor & ":" & PuertoDelServidor
```

File: CODIGO/ModLogin.bas [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+    For k = 1 To 100
+        serverOpt = RandomNumber(1, serverCount)
+    Next k
+    For k = 1 To 100
+        loginOpt = RandomNumber(1, loginServers)
+    Next k
+    IPdelServidor = ServerSettings.GetValue(environment, "ServerIp" & serverOpt)
+    PuertoDelServidor = ServerSettings.GetValue(environment, "PortPort" & serverOpt)
+    IPdelServidorLogin = ServerSettings.GetValue(environment, "LoginIp" & serverOpt)
+    PuertoDelServidorLogin = ServerSettings.GetValue(environment, "LoginPort" & serverOpt)
```
Comment (morgolock) [inline CODIGO/ModLogin.bas:103]:
```
Debug.Print "Using Login Server " & IPdelServidorLogin & ":" & PuertoDelServidorLogin
Debug.Print "Using Game Server " & IPdelServidor & ":" & PuertoDelServidor
```

File: CODIGO/ModLogin.bas [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+
+Public Sub DeleteCharRequestCode()
+    If UseBabelUI Then
+        Call RequestDeleteCode
+    Else
+        MsgBox ("Se ha enviado un código de verificación al mail proporcionado.")
+    End If
+End Sub
+
+Public Sub TransferChar(ByVal name As String, ByVal DestinationAccunt)
```
Comment (morgolock) [inline CODIGO/ModLogin.bas:300]: `DestinationAccount as string`

File: CODIGO/ModLogin.bas [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+    TransferCharNewOwner = DestinationAccunt
```
Comment (morgolock) [inline CODIGO/ModLogin.bas:301]:
```
debug.assert len( TransferCharNewOwner) > 0
debug.assert len(name)>0
```

File: CODIGO/Protocol.bas [hunk 10]
```diff
@@ -5773,28 +5776,13 @@ HandleRestOK_Err:
...
-    'Last Modification: 05/17/06
-    '
-    '***************************************************
-    
     On Error GoTo ErrHandler
-    
-    Call MsgBox(Reader.ReadString8())
-    
+    GetRemoteError = True
+    Call BabelUI.DisplayError(Reader.ReadString8(), "")
```
Comment (morgolock) [inline CODIGO/Protocol.bas:5782]: `if UseBabelUI Then     Call BabelUI.DisplayError(Reader.ReadString8(), "")` ?

File: CODIGO/clsBatch.cls [hunk 11]
```diff
@@ -164,21 +164,17 @@ Public Sub SetTexture(ByRef Texture As Direct3DTexture8)
     
     On Error GoTo SetTexture_Err
```
Comment (morgolock) [inline CODIGO/clsBatch.cls:166]: `debug.assert not texture is nothing`

File: CODIGO/clsTexManager.cls [hunk 12]
```diff
@@ -216,6 +216,54 @@ errhandler:
 
 End Function
 
+Public Function CreateTexture(ByVal Width As Long, ByVal Height As Long)
```
Comment (morgolock) [inline CODIGO/clsTexManager.cls:219]: `as Direct3DTexture8`

File: CODIGO/clsTexManager.cls [hunk 13]
```diff
@@ -216,6 +216,54 @@ errhandler:
...
+    Set CreateTexture = Texture
+    Exit Function
+ErrHandler:
+    Debug.Print "Failed to generate texture, " & Err.Description
+End Function
+
+Public Sub SetTextureData(ByRef Texture As Direct3DTexture8, ByRef Bytes() As Byte, ByVal size As Long, _
+                          ByVal TextureWidth As Long, ByVal DrawWidth As Integer, _
+                          ByVal startY As Integer, ByVal endY As Integer)
+On Error GoTo ErrHandler
```
Comment (morgolock) [inline CODIGO/clsTexManager.cls:232]: `debug.assert not texture is nothing`

File: CODIGO/clsTexManager.cls [hunk 14]
```diff
@@ -216,6 +216,54 @@ errhandler:
...
+    RenderArea.Top = startY
+    RenderArea.Bottom = endY
+    RenderArea.Right = DrawWidth
+    Call Texture.LockRect(0, lr, ByVal RenderArea, 0)
+    Dim BufferPos As Long
+    Dim row As Long
+    Dim destPtr As Long
+    destPtr = lr.pBits
+    For row = 0 To endY
+        Call DXCopyMemory(ByVal destPtr, Bytes(BufferPos), DrawWidth * 4)
```
Comment (morgolock) [inline CODIGO/clsTexManager.cls:245]: replace `4` with a const integer BytesPerPixel ?

File: CODIGO/clsTexManager.cls [hunk 15]
```diff
@@ -216,6 +216,54 @@ errhandler:
...
+        BufferPos = BufferPos + DrawWidth * 4
+        destPtr = destPtr + lr.Pitch
+    Next row
+    'Call DXCopyMemory(ByVal lr.pBits, Bytes(0), Size)
```
Comment (morgolock) [inline CODIGO/clsTexManager.cls:249]: remove line

File: CODIGO/clsTexManager.cls [hunk 16]
```diff
@@ -216,6 +216,54 @@ errhandler:
...
+   Exit Sub
+ErrHandler:
+    Call RegistrarError(Err.Number, Err.Description, "clsTexManager.SetTextureData", Erl)
+End Sub
+
+Public Function CreateTextureFromData(ByRef Bytes() As Byte, ByVal Size As Long, ByVal Width As Long, ByVal Height As Long) As Direct3DTexture8
+On Error GoTo ErrHandler
+    Dim Texture As Direct3DTexture8
+    Set Texture = mD3D.CreateTextureFromFileInMemoryEx(device, Bytes(0), UBound(Bytes) + 1, Width, Height, 1, 0, D3DFMT_A8R8G8B8, D3DPOOL_DEFAULT, D3DX_FILTER_LINEAR, D3DX_FILTER_LINEAR, &HFF000000, ByVal 0, ByVal 0)
+    Set CreateTextureFromData = Texture
```
Comment (morgolock) [inline CODIGO/clsTexManager.cls:260]: `debug.assert not texture is nothing`

**Review by morgolock — CHANGES_REQUESTED** [review]

File: CODIGO/General.bas (hunk 5 above)
Comment (matiascalegaris) [reply CODIGO/General.bas:1137]: adding log error inside Public Sub InitializeUI(ByVal Width As Long, ByVal Height As Long, ByVal pixelSize As Long)

File: CODIGO/clsTexManager.cls (hunk 16 above)
Comment (matiascalegaris) [reply CODIGO/clsTexManager.cls:260]: I'll place it but I'm not sure this is correct, this should be handled by the user of the function

Comment (morgolock) [reply CODIGO/clsTexManager.cls:260]: It's more to make our lives easier, if this is called with Texture nothing it should be clear from the assert that it's a mistake on the user

**Review by morgolock — APPROVED** [review]

**Review by morgolock — APPROVED** [review]

**Pull request merged**